src/Orchard.Web/Modules/Orchard.OpenId/InMemoryCache.cs

```diff
@@ -0,0 +1,62 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.IdentityModel.Clients.ActiveDirectory;
+
+namespace Orchard.OpenId {
+    public class InMemoryCache : TokenCache {
```

Comment:
I looked at the documentation for the token cache. And I think it would be much better if you actually had it registered as a singleton service such that each tenant would have its own list of tokens (not the case right now), and that this list would then be cleared when a tenant is restarted (not the case right now). Also, you would then not need to instantiate a new `InMemoryCache` instance for each request. Also, it needs to be a service such that when we use multiple nodes, we can use a token cache based on Redis for instance. And this service should be tied to the Azure feature.